File: commitment/canonical.go
package commitment

import (
	"fmt"
	"regexp"
	"strings"
	"time"

	"golang.org/x/text/cases"
	"golang.org/x/text/language"
	"golang.org/x/text/unicode/norm"
)

// jsSpace is ECMAScript's \s character class, spelled out.
//
// Go's own \s is ASCII-only ([\t\n\f\r ]) while JavaScript's includes NBSP,
// the Unicode space separators, the line/paragraph separators, and the BOM.
// Using Go's default would make the two implementations disagree the moment a
// form contained a non-breaking space — pasted out of a spreadsheet, say —
// and the disagreement would surface as an unverifiable document rather than
// as an error.
const jsSpace = `\t\n\v\f\r \x{00a0}\x{1680}\x{2000}-\x{200a}\x{2028}\x{2029}\x{202f}\x{205f}\x{3000}\x{feff}`

var (
	identifierNoise = regexp.MustCompile(`[` + jsSpace + `\-_/.]+`)
	whitespaceRun   = regexp.MustCompile(`[` + jsSpace + `]+`)

	// RFC 3339 with a mandatory offset, matching the TypeScript side exactly
	// (including the tolerated space separator and lowercase t/z).
	rfc3339 = regexp.MustCompile(
		`^(\d{4})-(\d{2})-(\d{2})[Tt ](\d{2}):(\d{2}):(\d{2})(\.\d+)?([Zz]|[+-]\d{2}:\d{2})$`)

	// Full Unicode uppercasing, language-neutral. Matches JavaScript's
	// toUpperCase, which does full case mapping (ß becomes SS); Go's
	// strings.ToUpper does simple per-rune mapping and would leave it alone.
	upper = cases.Upper(language.Und)
)

// Error is a canonicalization failure naming the offending field.
type Error struct {
	Field   string
	Message string
}

func (e *Error) Error() string { return e.Field + ": " + e.Message }

func errf(field, format string, args ...any) *Error {
	return &Error{Field: field, Message: fmt.Sprintf(format, args...)}
}

// NormalizeIdentifier applies NFC, strips separators, and uppercases.
func NormalizeIdentifier(v string) string {
	return upper.String(identifierNoise.ReplaceAllString(norm.NFC.String(v), ""))
}

// NormalizeText applies NFC, collapses internal whitespace runs to a single
// space, and trims.
//
// Trimming plain ASCII spaces is sufficient: the collapse has already
// rewritten every exotic space to U+0020.
func NormalizeText(v string) string {
	return strings.Trim(whitespaceRun.ReplaceAllString(norm.NFC.String(v), " "), " ")
}

// NormalizeTimestamp forces RFC 3339 input to UTC at second precision.
func NormalizeTimestamp(field, v string) (string, error) {
	trimmed := strings.TrimSpace(v)
	m := rfc3339.FindStringSubmatch(trimmed)
	if m == nil {
		return "", errf(field, "expected RFC 3339 with a UTC offset, got %q", v)
	}

	// Normalize the tolerated spellings before handing to the strict parser.
	normalized := trimmed
	if normalized[10] != 'T' {
		normalized = normalized[:10] + "T" + normalized[11:]
	}
	normalized = strings.Replace(normalized, "z", "Z", 1)

	// Go's parser validates the calendar date and rejects a seconds value of
	// 60, so impossible dates and leap seconds both fail here rather than
	// silently rolling forward into a date nobody typed.
	t, err := time.Parse(time.RFC3339, normalized)
	if err != nil {
		return "", errf(field, "not a valid date: %q", v)
	}

	return t.UTC().Truncate(time.Second).Format("2006-01-02T15:04:05Z"), nil
}

// NormalizeEnum validates a value against a closed set.
func NormalizeEnum(spec FieldSpec, v string) (string, error) {
	u := upper.String(strings.TrimSpace(norm.NFC.String(v)))
	for _, allowed := range spec.Values {
		if u == allowed {
			return u, nil
		}
	}
	return "", errf(spec.Name, "expected one of %s, got %q",
		strings.Join(spec.Values, ", "), v)
}

// CanonicalizeField canonicalizes a single raw value.
//
// Raw values arrive as nil, string, or an integer type. Floats are refused
// outright: money is integer cents, and rounding a caller's number silently
// would commit to a figure they never wrote.
func CanonicalizeField(spec FieldSpec, raw any) (Value, error) {
	if raw == nil {
		return nil, nil
	}

	if spec.Kind == KindInteger {
		switch n := raw.(type) {
		case int:
			return int64(n), nil
		case int64:
			return n, nil
		case float64:
			if n != float64(int64(n)) {
				return nil, errf(spec.Name,
					"expected an integer, got %v — money is integer cents, never a float", n)
			}
			return int64(n), nil
		default:
			return nil, errf(spec.Name, "expected a number, got %T", raw)
		}
	}

	s, ok := raw.(string)
	if !ok {
		return nil, errf(spec.Name, "expected a string for this field, got %T", raw)
	}

	switch spec.Kind {
	case KindIdentifier:
		return NormalizeIdentifier(s), nil
	case KindText:
		return NormalizeText(s), nil
	case KindTimestamp:
		return NormalizeTimestamp(spec.Name, s)
	case KindEnum:
		return NormalizeEnum(spec, s)
	}
	return nil, errf(spec.Name, "unhandled field kind")
}

// CanonicalizeForm canonicalizes a raw form into the fixed field set.
//
// Every field is always present in the result, explicitly nil when absent, so
// the tree has a constant shape for every form ever issued. Unknown keys are
// rejected rather than ignored: silently dropping a field the caller believed
// they were committing to is the worst available failure.
func CanonicalizeForm(raw map[string]any) (Form, error) {
	for k := range raw {
		if IndexOf(k) < 0 {
			return nil, errf(k, "not part of the committed field set (v%d)", FieldSetVersion)
		}
	}

	out := make(Form, len(Fields))
	for _, spec := range Fields {
		v, err := CanonicalizeField(spec, raw[spec.Name])
		if err != nil {
			return nil, err
		}
		out[spec.Name] = v
	}
	return out, nil
}

File: core/src/canonical.ts
/**
 * Canonicalization (§4.1).
 *
 * Every rule here is idempotent: canonicalize(canonicalize(x)) === canonicalize(x).
 * That property is tested, because a non-idempotent rule silently produces two
 * different roots for the same form depending on how many times it was
 * normalized on the way in.
 */

import {
  FIELDS,
  FIELD_SET_VERSION,
  type CanonicalForm,
  type FieldSpec,
  type RawForm,
} from './fields.js'

export class CanonicalizationError extends Error {
  constructor(
    readonly field: string,
    message: string,
  ) {
    super(`${field}: ${message}`)
    this.name = 'CanonicalizationError'
  }
}

/** Separator characters stripped from identifiers: whitespace - _ / . */
const IDENTIFIER_NOISE = /[\s\-_/.]+/g

/**
 * RFC 3339 with a mandatory offset. The offset is what makes this a point in
 * time rather than a wall-clock reading, so a naive datetime is an error, not
 * something to guess a zone for.
 */
const RFC3339 =
  /^(\d{4})-(\d{2})-(\d{2})[Tt ](\d{2}):(\d{2}):(\d{2})(\.\d+)?([Zz]|[+-]\d{2}:\d{2})$/

export function normalizeIdentifier(value: string): string {
  return value.normalize('NFC').replace(IDENTIFIER_NOISE, '').toUpperCase()
}

export function normalizeText(value: string): string {
  return value.normalize('NFC').replace(/\s+/g, ' ').trim()
}

/**
 * Truncates to second precision rather than rejecting sub-second input — an
 * 8130-3 is not signed to the millisecond, and rejecting would make round
 * trips through systems that add milliseconds fail confusingly.
 */
export function normalizeTimestamp(field: string, value: string): string {
  const trimmed = value.trim()
  const m = RFC3339.exec(trimmed)
  if (!m) {
    throw new CanonicalizationError(
      field,
      `expected RFC 3339 with a UTC offset, got ${JSON.stringify(value)}`,
    )
  }

  const [, y, mo, d, hh, mm, ss] = m
  const year = Number(y)
  const month = Number(mo)
  const day = Number(d)

  // Date.parse rolls impossible dates forward — 2019-02-30 silently becomes
  // 2019-03-02 — which would commit to a date nobody typed. Validate the
  // calendar date by round-tripping it through a UTC construction instead.
  const probe = new Date(Date.UTC(year, month - 1, day))
  if (
    probe.getUTCFullYear() !== year ||
    probe.getUTCMonth() !== month - 1 ||
    probe.getUTCDate() !== day
  ) {
    throw new CanonicalizationError(
      field,
      `not a real calendar date: ${JSON.stringify(value)}`,
    )
  }

  // Leap seconds are rejected rather than rolled forward, for the same reason.
  if (Number(hh) > 23 || Number(mm) > 59 || Number(ss) > 59) {
    throw new CanonicalizationError(
      field,
      `time out of range: ${JSON.stringify(value)}`,
    )
  }

  const ms = Date.parse(trimmed)
  if (Number.isNaN(ms)) {
    throw new CanonicalizationError(field, `not a valid date: ${JSON.stringify(value)}`)
  }
  return new Date(Math.floor(ms / 1000) * 1000).toISOString().replace(/\.\d{3}Z$/, 'Z')
}

export function normalizeInteger(field: string, value: unknown): number {
  if (typeof value !== 'number') {
    throw new CanonicalizationError(
      field,
      `expected a number, got ${typeof value} — integers are never parsed from strings here`,
    )
  }
  if (!Number.isFinite(value)) {
    throw new CanonicalizationError(field, `expected a finite number, got ${value}`)
  }
  if (!Number.isSafeInteger(value)) {
    throw new CanonicalizationError(
      field,
      `expected a safe integer, got ${value} — money is integer cents, never a float`,
    )
  }
  return value
}

function normalizeEnum(spec: FieldSpec, value: string): string {
  const upper = value.normalize('NFC').trim().toUpperCase()
  if (!spec.values!.includes(upper)) {
    throw new CanonicalizationError(
      spec.name,
      `expected one of ${spec.values!.join(', ')}, got ${JSON.stringify(value)}`,
    )
  }
  return upper
}

/**
 * Canonicalizes one field value.
 *
 * `null` and absent are the same thing. An empty string is NOT null — a shop
 * that wrote nothing in the remarks box committed to an empty remarks box, and
 * that is a different claim from having no remarks field at all.
 */
export function canonicalizeField(
  spec: FieldSpec,
  value: unknown,
): string | number | null {
  if (value === null || value === undefined) return null

  if (spec.kind === 'integer') return normalizeInteger(spec.name, value)

  if (typeof value !== 'string') {
    throw new CanonicalizationError(
      spec.name,
      `expected a string for a ${spec.kind} field, got ${typeof value}`,
    )
  }

  switch (spec.kind) {
    case 'identifier':
      return normalizeIdentifier(value)
    case 'text':
      return normalizeText(value)
    case 'timestamp':
      return normalizeTimestamp(spec.name, value)
    case 'enum':
      return normalizeEnum(spec, value)
  }
}

/**
 * Canonicalizes a whole form into the fixed field set.
 *
 * Every field is always present in the result, explicitly null when absent, so
 * the commitment tree has a constant shape for every form ever issued. Unknown
 * keys are rejected rather than ignored: silently dropping a field the caller
 * believed they were committing to is the worst possible failure here.
 */
export function canonicalizeForm(form: RawForm): CanonicalForm {
  const unknown = Object.keys(form).filter(
    (k) => !FIELDS.some((f) => f.name === k),
  )
  if (unknown.length > 0) {
    throw new CanonicalizationError(
      unknown[0]!,
      `not part of the committed field set (v${FIELD_SET_VERSION}); ` +
        `got unknown keys: ${unknown.join(', ')}`,
    )
  }

  const out: CanonicalForm = {}
  for (const spec of FIELDS) {
    out[spec.name] = canonicalizeField(spec, form[spec.name])
  }
  return out
}
